src/DDTrace/Integrations/SymfonyMessenger/SymfonyMessengerIntegration.php

```diff
@@ -0,0 +1,348 @@
+<?php
+
+namespace DDTrace\Integrations\SymfonyMessenger;
+
+use DDTrace\HookData;
+use DDTrace\Integrations\Integration;
+use DDTrace\SpanData;
+use DDTrace\Tag;
+use DDTrace\Util\ObjectKVStore;
+use Symfony\Component\Messenger\Bridge\AmazonSqs\Transport\AmazonSqsReceivedStamp;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Handler\HandlerDescriptor;
+use Symfony\Component\Messenger\Handler\HandlersLocatorInterface;
+use Symfony\Component\Messenger\Stamp\BusNameStamp;
+use Symfony\Component\Messenger\Stamp\ConsumedByWorkerStamp;
+use Symfony\Component\Messenger\Stamp\DelayStamp;
+use Symfony\Component\Messenger\Stamp\HandledStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+use Symfony\Component\Messenger\Stamp\RedeliveryStamp;
+use Symfony\Component\Messenger\Stamp\SentStamp;
+use Symfony\Component\Messenger\Stamp\TransportMessageIdStamp;
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use function DDTrace\hook_method;
+use function DDTrace\install_hook;
+use function DDTrace\remove_hook;
+use function DDTrace\trace_method;
+
+class SymfonyMessengerIntegration extends Integration
+{
+    const NAME = 'symfonymessenger';
+
+    public function init(): int
+    {
+        $integration = $this;
+
+        if (!\class_exists(\Symfony\Component\Messenger\Event\WorkerStartedEvent::class)) {
+            // Only exists in Symfony Messenger 4.4+
+            return Integration::NOT_LOADED;
+        }
+
+        trace_method(
+            'Symfony\Component\Messenger\MessageBusInterface',
+            'dispatch',
+            function (SpanData $span, array $args, $retval, $exception) use ($integration) {
+                $integration->setSpanAttributes($span, 'symfony.messenger.dispatch', null, $retval ?? $args[0], null, null, true);
+                if ($exception) {
+                    // Worker::handleMessage() will catch the exception. We need to manually attach it to the root span.
+                    \DDTrace\root_span()->exception = $exception;
+                }
+            }
+        );
+
+        // Attach current context to Envelope before sender sends it to remote queue
+        install_hook(
+            'Symfony\Component\Messenger\Transport\Sender\SenderInterface::send',
+            function (HookData $hook) {
+                /** @var \Symfony\Component\Messenger\Envelope $envelope */
+                $envelope = $hook->args[0];
+
+                if (\ddtrace_config_distributed_tracing_enabled()) {
+                    $ddTraceStamp = $envelope->last(DDTraceStamp::class);
+
+                    // Add distributed tracing stamp only if not already on the envelope
+                    if ($ddTraceStamp === null) {
+                        $tracingHeaders = \DDTrace\generate_distributed_tracing_headers();
+                        $hook->overrideArguments([
+                            $envelope->with(new DDTraceStamp($tracingHeaders))
```

Comment:
This may be wrong. `Envelope` [is immutable](https://github.com/symfony/symfony/blob/9a0266029867201e5637af634eaa8a263433dc07/src/Symfony/Component/Messenger/Envelope.php#L57-L66), so this has to be done like this:
```suggestion
                            $envelope = $envelope->with(new DDTraceStamp($tracingHeaders))
```

The issue here is that you're not changing `$envelope` by reference, so we may be out of luck :(

Reply:
In this piece of code, we're _overriding_ the argument of `SenderInterface::send` (which only takes one argument: an `Envelope` object) by another `Envelope` object with a custom stamp containing the distributed trace context information. This is done using a custom method from our API (`\DDTrace\HookData::overrideArguments`). As a result of this function call, the newly provided envelope will be used by the method call instead of the original one.

Doing `$envelope = $envelope->with(new DDTraceStamp($tracingHeaders))` would effectively replace the envelope within the scope of the prehook, but not the actual method call :+1: